src/query/api/v1/handler/database/kvstore.go

```diff
@@ -0,0 +1,203 @@
+	nsproto "github.com/m3db/m3/src/dbnode/generated/proto/namespace"
+	"github.com/m3db/m3/src/dbnode/kvconfig"
+	"github.com/m3db/m3/src/query/util/logging"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	"github.com/m3db/m3/src/x/instrument"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+)
+
+const (
+	// KeyValueStoreURL is the url to edit key/value configuration values.
+	KeyValueStoreURL = "/kvstore"
+	// KeyValueStoreHTTPMethod is the HTTP method used with this resource.
+	KeyValueStoreHTTPMethod = http.MethodPost
+)
+
+// KeyValueUpdate defines an update to a key's value.
+type KeyValueUpdate struct {
+	// Key to update.
+	Key string `json:"key"`
+	// Value to update the key to.
+	Value json.RawMessage `json:"value"`
+	// Commit, if false, will not persist the update. If true, the
+	// update will be persisted. Used to test format of inputs.
+	Commit bool `json:"commit"`
+}
+
+// KeyValueUpdateResult defines the result of an update to a key's value.
+type KeyValueUpdateResult struct {
+	// Key to update.
+	Key string `json:"key"`
+	// Old is the value before the update.
+	Old string `json:"old"`
+	// New is the value after the update.
+	New string `json:"new"`
+	// Version of the key.
+	Version int `json:"version"`
+}
+
+// KeyValueStoreHandler represents a handler for the key/value store endpoint
+type KeyValueStoreHandler struct {
+	client         clusterclient.Client
+	instrumentOpts instrument.Options
+}
+
+// NewKeyValueStoreHandler returns a new instance of handler
+func NewKeyValueStoreHandler(
+	client clusterclient.Client,
+	instrumentOpts instrument.Options,
+) http.Handler {
+	return &KeyValueStoreHandler{
+		client:         client,
+		instrumentOpts: instrumentOpts,
+	}
+}
+
+func (h *KeyValueStoreHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	logger := logging.WithContext(r.Context(), h.instrumentOpts)
+
+	update, err := h.parseBody(r)
+	if err != nil {
+		logger.Error("unable to parse request", zap.Error(err))
+		xhttp.WriteError(w, err)
+		return
+	}
+
+	kvStore, err := h.client.KV()
+	if err != nil {
+		logger.Error("unable to get kv store", zap.Error(err))
+		xhttp.WriteError(w, err)
+		return
+	}
+
+	results, err := h.update(logger, kvStore, update)
+	if err != nil {
+		logger.Error("kv store error",
+			zap.Error(err),
+			zap.Any("update", update))
+		xhttp.WriteError(w, err)
+		return
+	}
+
+	xhttp.WriteJSONResponse(w, results, logger)
+}
+
+func (h *KeyValueStoreHandler) parseBody(r *http.Request) (*KeyValueUpdate, error) {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, xerrors.NewInvalidParamsError(err)
+	}
+	defer r.Body.Close()
+
+	var parsed KeyValueUpdate
+	if err := json.Unmarshal(body, &parsed); err != nil {
+		return nil, xerrors.NewInvalidParamsError(err)
+	}
+
+	return &parsed, nil
+}
+
+func (h *KeyValueStoreHandler) update(
+	logger *zap.Logger,
+	kvStore kv.Store,
+	update *KeyValueUpdate,
+) (*KeyValueUpdateResult, error) {
+	old, err := kvStore.Get(update.Key)
```

Comment:
Allowing arbitrary kv updates here is a little scary; seems the footgunning potential may be quite high, although forcing the Commit flag may be enough for safety

Reply:
Well so they have to conform to the type of input (e.g. the key for querylimits must be as `kvpb.QueryLimits{}`). And yeah I think the Commit allows you to safely confirm the diff which we return in the result before applying.

Reply:
Could we use `CheckAndSet` instead to require the client to know what version is being overridden?